src/file_pipe_log.rs

```diff
@@ -0,0 +1,959 @@
+    ) -> Result<(FileId, usize)> {
+        if let Some(content) = batch.encode_to_bytes(self.compression_threshold) {
+            // TODO: `pwrite` is performed in the mutex. Is it possible for concurrence?
+            let mut chunk = None;
+            let (file_id, offset, fd) = if queue == LogQueue::Append {
+                let mut cache_submitor = self.cache_submitor.lock().unwrap();
+                let (file_id, offset, fd) = self.append_bytes(queue, &content, &mut sync)?;
+                chunk = cache_submitor.submit_new_block(file_id, offset, batch.entries_size());
+                (file_id, offset, fd)
+            } else {
+                self.append_bytes(queue, &content, &mut sync)?
+            };
+            if sync {
+                fd.sync()?;
+            }
+
+            for item in batch.items.iter_mut() {
+                if let LogItemContent::Entries(entries) = &mut item.content {
+                    entries.set_position(queue, file_id, offset, &chunk);
+                }
+            }
+
+            return Ok((file_id, content.len()));
+        }
+        Ok((Default::default(), 0))
+    }
+
+    fn sync(&self, queue: LogQueue) -> Result<()> {
+        if let Some(fd) = self.get_queue(queue).get_active_fd() {
+            fd.sync()?;
+        }
+        Ok(())
+    }
+
+    fn active_file_id(&self, queue: LogQueue) -> FileId {
+        self.get_queue(queue).active_file_id
+    }
+
+    fn first_file_id(&self, queue: LogQueue) -> FileId {
+        self.get_queue(queue).first_file_id
+    }
+
+    fn file_at(&self, queue: LogQueue, position: f64) -> FileId {
+        // TODO: sanitize position
+        let cur_size = self.total_size(queue);
+        let count = (cur_size as f64 * position) as usize / self.rotate_size;
+        let file_num = self.get_queue(queue).first_file_id.forward(count);
+        assert!(file_num <= self.active_file_id(queue));
+        file_num
+    }
+
+    fn new_log_file(&self, queue: LogQueue) -> Result<()> {
+        self.mut_queue(queue).new_log_file()
+    }
+
+    fn purge_to(&self, queue: LogQueue, file_id: FileId) -> Result<usize> {
+        let (mut manager, name_suffix) = match queue {
+            LogQueue::Append => (self.appender.wl(), LOG_SUFFIX),
+            LogQueue::Rewrite => (self.rewriter.wl(), REWRITE_SUFFIX),
+        };
+        let purge_count = manager.purge_to(file_id)?;
+        drop(manager);
+
+        let mut cur_file_id = file_id.backward(purge_count);
+        for i in 0..purge_count {
+            let mut path = PathBuf::from(&self.dir);
+            path.push(generate_file_name(cur_file_id, name_suffix));
+            if let Err(e) = fs::remove_file(&path) {
+                warn!("Remove purged log file {:?} fail: {}", path, e);
+                return Ok(i);
+            }
+            cur_file_id = cur_file_id.forward(1);
+        }
+        Ok(purge_count)
+    }
+}
+
+fn generate_file_name(file_num: FileId, suffix: &'static str) -> String {
```

Comment:
Maybe replace `file_num` with `file_id` is better to be consistent with the other interfaces.